# Coding Exercises (Appendix 3)

## Exercise 8: N-Dimensional Numpy Arrays

In [1]:
# run the cell!
import numpy as np
np.set_printoptions(precision=2, suppress= True)

Let´s assume we have three investment projects with the following cashflow plans:

In [2]:
# run the cell!
project_1 = [-500, 100, 150, 200, 200, 175, 100, 50]
project_2 = [-750, 100, 125, 150, 175, 200, 250, 0]
project_3 = [-2000, 500, 1000, 2500, 0, 0, 0, 0]

1. Create and save a __2-dimensional Numpy Array__ (in the variable __projects__).

In [3]:
projects = np.array([project_1, project_2, project_3])

In [4]:
projects

array([[ -500,   100,   150,   200,   200,   175,   100,    50],
       [ -750,   100,   125,   150,   175,   200,   250,     0],
       [-2000,   500,  1000,  2500,     0,     0,     0,     0]])

2. Check the __shape__ of projects. __How many rows/columns__ do we have?

In [5]:
projects.shape

(3, 8)

(result: 3 rows, 8 columns)

3. __Slice__ projects for __the second project__.

In [6]:
projects[1]

array([-750,  100,  125,  150,  175,  200,  250,    0])

4. Return the __initial Investment__ of __Project 3__.

In [7]:
projects[-1, 0]

-2000

(result: -2000)

5. Return the __final cashflow__ in year 7 for __all projects__.

In [8]:
projects[:, -1]

array([50,  0,  0])

6. Return the first positive cashflows (__Year 1__) for __Projects 1 and 2__.

In [9]:
projects[:2, 1]

array([100, 100])

(result: array([100, 100]))

7. __Change__ the initial investment for __project 3__ to __-1900__.

In [10]:
projects[-1, 0] = -1900

In [11]:
projects

array([[ -500,   100,   150,   200,   200,   175,   100,    50],
       [ -750,   100,   125,   150,   175,   200,   250,     0],
       [-1900,   500,  1000,  2500,     0,     0,     0,     0]])

8. Change the __last cashflows__ in year 7 for all projects to __25__.

In [12]:
projects[:, -1] = 25

In [13]:
projects

array([[ -500,   100,   150,   200,   200,   175,   100,    25],
       [ -750,   100,   125,   150,   175,   200,   250,    25],
       [-1900,   500,  1000,  2500,     0,     0,     0,    25]])

9. Calculate the total cashflows on each timestamp over all three projects (__column-wise sum__).

In [14]:
projects.sum(axis = 0)


array([-3150,   700,  1275,  2850,   375,   375,   350,    75])

(results: array([-3150,   700,  1275,  2850,   375,   375,   350,    75]))

10. Calculate __cumulative cashflows__ for all three projects.

In [15]:
projects.cumsum(axis=1)

array([[ -500,  -400,  -250,   -50,   150,   325,   425,   450],
       [ -750,  -650,  -525,  -375,  -200,     0,   250,   275],
       [-1900, -1400,  -400,  2100,  2100,  2100,  2100,  2125]],
      dtype=int32)

11. Convert __all projects__ to EUR currency (__multiply with 0.9__).

In [16]:
projects * 0.9

array([[ -450. ,    90. ,   135. ,   180. ,   180. ,   157.5,    90. ,
           22.5],
       [ -675. ,    90. ,   112.5,   135. ,   157.5,   180. ,   225. ,
           22.5],
       [-1710. ,   450. ,   900. ,  2250. ,     0. ,     0. ,     0. ,
           22.5]])

# Well Done!

---------------------------------------------------------------------------------------------------------------------